### PR TITLE
Add Unity Oculus VR + Codemagic build & publish guide

### PR DESCRIPTION
**Pull request by MinaPecheux**
Hello,

here is a PR to add a new doc page (in the YAML quick start section) on how to setup a Codemagic workflow to build and publish a Unity VR app for Oculus. It describes how to configure the Codemagic app environment variables, the Unity project settings and the codemagic.yaml file to distribute a VR app on an Oculus release channel using the Oculus CLI tools.

Some parts are copied from [the other Unity guide in the doc](docs.codemagic.io/yaml-quick-start/building-a-unity-app/), I hope it's fine.

(Based of my previous article on Codemagic's blog: https://blog.codemagic.io/building-and-publishing-a-vr-unity-game-for-oculus-with-codemagic/)

Cheers! :)

### CODE REVIEW

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+1. On the Applications page in Codemagic, click the **Add application** button and follow the on-screen instructions to add your Unity project.
+2. Add environment variables for Unity as described [here](#unity-variables).
+3. Add environment variables for Android code signing as described [here](#android-variables).
+4. Add environment variables for Oculus app signing as described [here](#oculus-variables).
+5. Add a build script to Unity as explained [here](#unity-build-script).
+6. Configure the Android and VR build settings in Unity as described [here](#unity-build-settings).
+7. Configure Custom Gradle templates as described [here](#custom-gradle-template).
+8. Create a codemagic.yaml workflow configuration using the Unity template and configure with your details as shown [here](#workflow-configuration).
+
+<h2 id="unity-variables">Environment variables for Unity</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:37]:
```suggestion
## Environment variables for Unity {#unity-variables}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+1. Create a variable called `UNITY_SERIAL` and set the value to your Unity serial number. In the **Select group** dropdown type `unity` and click on the **Create "unity" group** button. Mark the variable as **secure** to encrypt the value and click the **Add** button.
+2. Create a variable called `UNITY_USERNAME` and set the value to **email address** used with your Unity ID, add it to the "unity" group, mark this as **secure** to encrypt the value and click the **Add** button.
+3. Create a variable called `UNITY_PASSWORD` and set the value to your Unity ID **password**, add to the "unity" group, mark this as **secure** to encrypt the value and click the **Add** button.
+
+Note that the environment variable `UNITY_HOME` is already set on the build machines.
+
+On the macOS Unity base image `UNITY_HOME` is set to `/Applications/Unity/Hub/Editor/2020.3.28f1/Unity.app`.
+
+<h2 id="android-variables">Environment variables for Android code signing</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:55]:
```suggestion
## Environment variables for Android code signing {#android-variables}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+You will need to set the following environment variables in a variable group called `keystore_credentials` for Android code signing:
+
+`CM_KEYSTORE_PATH`, `CM_KEYSTORE`, `CM_KEYSTORE_PASSWORD`, `CM_KEY_PASSWORD` and `CM_KEY_ALIAS`.
+
+Please refer to the documentation about signing Android apps [here](https://docs.codemagic.io/yaml-code-signing/signing-android/) for further details.
+
+<h2 id="oculus-variables">Environment variables for Oculus app signing</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:63]:
```suggestion
## Environment variables for Oculus app signing {#oculus-variables}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+You will need to set the following environment variables in a variable group called `oculus` for Oculus app signing:
+
+`OCULUS_APP_ID` and `OCULUS_APP_SECRET` (or `OCULUS_USER_TOKEN`)
+
+You can find this info in your Oculus app dashboard, in the API section:
+
+- The `OCULUS_APP_ID` is the **App ID** field of your app.
+- If you own the application, then you'll be able to get the **App secret** field of your app and use it to fill the `OCULUS_APP_SECRET` Codemagic environment variable. Else, you will need to generate a user token and add it to your Codemagic environment variable as `OCULUS_USER_TOKEN`.
+
+<h2 id="unity-build-script">Add a build script to Unity</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:74]:
```suggestion
## Add a build script to Unity {#unity-build-script}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+        private static string[] GetScenes()
+        {
+            return (from scene in EditorBuildSettings.scenes where scene.enabled select scene.path).ToArray();
+        }
+
+    }
+    ```
+
+<h2 id="unity-build-settings">Configure Android and VR build settings in Unity</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:226]:
```suggestion
## Configure Android and VR build settings in Unity {#unity-build-settings}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+1. Click on the XR Plug-in Management tab and pick the **Oculus** option for the Android build.
+2. Click on the Player settings tab.
+3. Expand 'Other Settings' and check the 'Override Default Package Name' checkbox.
+4. Enter the package name for your app, e.g. com.domain.yourappname.
+5. Set the 'Version number'.
+6. Set the 'Minimum API Level' and 'Target API Level' to **Android 10.0 (API Level 29)** (be careful, higher versions are not supported by Oculus at the moment).
+7. In the 'Configuration' section set 'Scripting Backend' to **IL2CPP**.
+
+<h2 id="custom-gradle-template">Add a custom base Gradle template</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:242]:
```suggestion
## Add a custom base Gradle template {#custom-gradle-template}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        }
+    }
+}
+
+task clean(type: Delete) {
+    delete rootProject.buildDir
+}
+```
+
+<h2 id="workflow-configuration">Workflow configuration with codemagic.yaml</h2>
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:293]:
```suggestion
## Workflow configuration with codemagic.yaml {#workflow-configuration}
```

File: content/yaml-quick-start/building-a-vr-oculus-app.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+    private static string[] GetScenes()
+    {
+        return (from scene in EditorBuildSettings.scenes where scene.enabled select scene.path).ToArray();
+    }
+
+}
+```
+
+**Warning:** this script does not auto-increment the bundle version of your Android app so you will need to:
```
Comment (nevercode-kevin) [inline content/yaml-quick-start/building-a-vr-oculus-app.md:164]: Why not use the workflow build number (BUILD_NUMBER)  or project build number (PROJECT_BUILD_NUMBER) from Codemagic to increment `bundleVersionCode`? 

Does Oculus expect a different `bundleVersion` for each submission, or can that remain the same but with a different `bundleVersionCode`?

I would have thought you could just do this so no additional C# scripts have to be added:

```        // Set bundle version
        var versionIsSet = int.TryParse(Environment.GetEnvironmentVariable("BUILD_NUMBER"), out int version);
        if (versionIsSet)
        {
            Debug.Log($"Bundle version code set to {version}");
            PlayerSettings.Android.bundleVersionCode = version;
        }
        else
        {
            Debug.Log("Bundle version not provided");
        }
```

**Review by nevercode-kevin — CHANGES_REQUESTED** [review]

File: content/yaml-quick-start/building-a-vr-oculus-app.md (hunk 8 above)
Comment (MinaPecheux) [reply content/yaml-quick-start/building-a-vr-oculus-app.md:164]: Hi @nevercode-kevin ! So yes it needs to be different each time, just like when you publish on the Play Store for example... but you're right, we can just increase the last code digit I guess. The important thing is that the full code (eg. 1.0.2 changes each time).
(And contrary to Play Store, the Oculus CLI tool has no command to get the version of the latest build sadly :/)

Thanks for the suggestion, I'll update the docs! :)

**Comment by MinaPecheux** [comment]
@nevercode-kevin I've updated the doc to use your auto-versioning based on the Codemagic workflow index :)

File: content/yaml-quick-start/building-a-vr-oculus-app.md (hunk 8 above)
Comment (nevercode-kevin) [reply content/yaml-quick-start/building-a-vr-oculus-app.md]: Just to clarify, what does Oculus expect? For example, is it ok to have version 1.0.0 and then build 1, 2, 3 of the same version? Or can it have the same build number but a different version e.g. version 1.0.1 (build number 1), version 1.0.2 (build number 1)?

Comment (MinaPecheux) [reply content/yaml-quick-start/building-a-vr-oculus-app.md]: Hey! So, according to these docs: https://developer.oculus.com/resources/publish-mobile-manifest/#unity
It looks like the important part is indeed to increase the **Bundle Version Code** and not necessarily the entire **Bundle Version**.
So I'm guessing something like `1.0.2b0 -> 1.0.2.b1` would be fine :)
(but please tell me if I'm not reading their docs right and you feel it won't work!)

Comment (nevercode-kevin) [reply content/yaml-quick-start/building-a-vr-oculus-app.md]: Yes, that would be fine.

**Review by nevercode-kevin — APPROVED** [review]
LGTM

**Pull request merged**